Incorrect caret position in empty elements with a vertical writing mode

pxlcoder opened the pull request:
#### 3931fcef84e408738e94e3a8dea06f611a68b6ec
<pre>
Incorrect caret position in empty elements with a vertical writing mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=248322">https://bugs.webkit.org/show_bug.cgi?id=248322</a>
<a href="https://rdar.apple.com/102652127">rdar://102652127</a>

Reviewed by Tim Nguyen.

Caret rect computation for empty elements does not fully account for vertical
writing mode. Fix by consistently using logical values.

* LayoutTests/editing/selection/caret-in-empty-block-vertical-horizontal-rtl-text-indent-padding-expected.txt: Added.
* LayoutTests/editing/selection/caret-in-empty-block-vertical-horizontal-rtl-text-indent-padding.html: Added.
* LayoutTests/editing/selection/caret-in-empty-block-vertical-rl-rtl-expected.txt: Added.
* LayoutTests/editing/selection/caret-in-empty-block-vertical-rl-rtl.html: Added.
* LayoutTests/editing/selection/caret-in-empty-inline-vertical-rl-rtl-expected.txt: Added.
* LayoutTests/editing/selection/caret-in-empty-inline-vertical-rl-rtl.html: Added.
* Source/WebCore/rendering/CaretRectComputation.cpp:
(WebCore::computeCaretRectForEmptyElement):
(WebCore::computeCaretRectForBlock):
(WebCore::computeCaretRectForInline):
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
(WebCore::RenderBoxModelObject::borderAndPaddingLogicalHeight const):
(WebCore::RenderBoxModelObject::borderAndPaddingLogicalLeft const):
(WebCore::RenderBoxModelObject::borderAndPaddingLogicalRight const):
(WebCore::RenderBoxModelObject::borderAndPaddingEnd const):

Canonical link: <a href="https://commits.webkit.org/270707@main">https://commits.webkit.org/270707@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f41d6a49408a83f778ff4feeb2ccc94e0706786

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4736 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27399 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28222 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23921 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26451 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6535 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2148 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23961 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26381 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3616 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22486 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28798 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3222 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23446 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29497 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23837 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23832 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27388 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3241 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1439 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4646 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23194 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6296 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3707 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/CaretRectComputation.cpp:
```diff
@@ -91,8 +91,8 @@ static LayoutRect computeCaretRectForEmptyElement(const RenderBoxModelObject& re
-    LayoutUnit x = renderer.borderLeft() + renderer.paddingLeft();
-    LayoutUnit maxX = width - renderer.borderRight() - renderer.paddingRight();
+    LayoutUnit x = renderer.borderAndPaddingStart();
+    LayoutUnit maxX = logicalWidth - renderer.borderAndPaddingEnd();
```
have you tested this change with RTL/LTR + text indent? It looks like now x/maxX is logical but some of the code below (alignment) still thinks it is not.

pxlcoder replied:
Will fix.

pxlcoder replied:
Fixed and added another test.

pxlcoder committed:
Incorrect caret position in empty elements with a vertical writing mode
https://bugs.webkit.org/show_bug.cgi?id=248322
rdar://102652127

Reviewed by Tim Nguyen.

Caret rect computation for empty elements does not fully account for vertical
writing mode. Fix by consistently using logical values.

* LayoutTests/editing/selection/caret-in-empty-block-vertical-horizontal-rtl-text-indent-padding-expected.txt: Added.
* LayoutTests/editing/selection/caret-in-empty-block-vertical-horizontal-rtl-text-indent-padding.html: Added.
* LayoutTests/editing/selection/caret-in-empty-block-vertical-rl-rtl-expected.txt: Added.
* LayoutTests/editing/selection/caret-in-empty-block-vertical-rl-rtl.html: Added.
* LayoutTests/editing/selection/caret-in-empty-inline-vertical-rl-rtl-expected.txt: Added.
* LayoutTests/editing/selection/caret-in-empty-inline-vertical-rl-rtl.html: Added.
* Source/WebCore/rendering/CaretRectComputation.cpp:
(WebCore::computeCaretRectForEmptyElement):
(WebCore::computeCaretRectForBlock):
(WebCore::computeCaretRectForInline):
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
(WebCore::RenderBoxModelObject::borderAndPaddingLogicalHeight const):
(WebCore::RenderBoxModelObject::borderAndPaddingLogicalLeft const):
(WebCore::RenderBoxModelObject::borderAndPaddingLogicalRight const):
(WebCore::RenderBoxModelObject::borderAndPaddingEnd const):

Canonical link: https://commits.webkit.org/270707@main
